src/python/zquantum/core/circuits/_operations.py

```diff
@@ -0,0 +1,83 @@
+from functools import singledispatch
+from numbers import Number
+from typing import Tuple, Union, Dict, TypeVar, Iterable
+
+import sympy
+from typing_extensions import Protocol
+
+Parameter = Union[sympy.Symbol, Number]
+
+T = TypeVar("T", bound="Operation")
+
+
+class Operation(Protocol):
+    """Represents arbitrary operation applicable to a circuit or wavefunction."""
+
+    @property
+    def params(self) -> Tuple[Parameter, ...]:
+        """Parameters of this operation."""
+        raise NotImplementedError()
```

Comment:
WDYT about decorating the protocol's methods with `@abstractmethod`? This would allow opt-in check for implementers. We could also remove the `raise NotImplementedError` lines, making codecov happy at the same time.